homeassistant/components/pi_hole/__init__.py

```diff
@@ -16,95 +16,172 @@
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema(vol.All([PI_HOLE_SCHEMA], ensure_unique_names))},
     extra=vol.ALLOW_EXTRA,
 )
 
 SERVICE_DISABLE_SCHEMA = vol.Schema(
     {
         vol.Required(SERVICE_DISABLE_ATTR_DURATION): vol.All(
             cv.time_period_str, cv.positive_timedelta
-        )
+        ),
+        vol.Optional(SERVICE_DISABLE_ATTR_NAME): cv.string,
     }
 )
 
+SERVICE_ENABLE_SCHEMA = vol.Schema({vol.Optional(SERVICE_ENABLE_ATTR_NAME): cv.string})
+
 
 async def async_setup(hass, config):
     """Set up the pi_hole integration."""
+    hass.data[DOMAIN] = {}
+
+    for conf in config[DOMAIN]:
+        name = conf[CONF_NAME]
+        host = conf[CONF_HOST]
+        use_tls = conf[CONF_SSL]
+        verify_tls = conf[CONF_VERIFY_SSL]
+        location = conf[CONF_LOCATION]
+        api_key = conf.get(CONF_API_KEY)
+
+        LOGGER.debug("Setting up %s integration with host %s", DOMAIN, host)
+
+        session = async_get_clientsession(hass, verify_tls)
+        pi_hole = PiHoleData(
+            Hole(
+                host,
+                hass.loop,
+                session,
+                location=location,
+                tls=use_tls,
+                api_token=api_key,
+            ),
+            name,
+        )
 
-    conf = config[DOMAIN]
-    name = conf[CONF_NAME]
-    host = conf[CONF_HOST]
-    use_tls = conf[CONF_SSL]
-    verify_tls = conf[CONF_VERIFY_SSL]
-    location = conf[CONF_LOCATION]
-    api_key = conf.get(CONF_API_KEY)
-
-    LOGGER.debug("Setting up %s integration with host %s", DOMAIN, host)
-
-    session = async_get_clientsession(hass, verify_tls)
-    pi_hole = PiHoleData(
-        Hole(
-            host, hass.loop, session, location=location, tls=use_tls, api_token=api_key
-        ),
-        name,
-    )
-
-    await pi_hole.async_update()
-
-    hass.data[DOMAIN] = pi_hole
+        await pi_hole.async_update()
 
-    async def handle_disable(call):
-        if api_key is None:
-            raise vol.Invalid("Pi-hole api_key must be provided in configuration")
+        hass.data[DOMAIN][slugify(name)] = pi_hole
 
+    async def disable_service_handler(call):
+        """Handle the service call to disable a single Pi-Hole or all configured Pi-Holes."""
         duration = call.data[SERVICE_DISABLE_ATTR_DURATION].total_seconds()
-
-        LOGGER.debug("Disabling %s %s for %d seconds", DOMAIN, host, duration)
-        await pi_hole.api.disable(duration)
-
-    async def handle_enable(call):
-        if api_key is None:
-            raise vol.Invalid("Pi-hole api_key must be provided in configuration")
-
-        LOGGER.debug("Enabling %s %s", DOMAIN, host)
-        await pi_hole.api.enable()
+        name = call.data.get(SERVICE_DISABLE_ATTR_NAME)
+
+        async def do_disable(name):
+            """Disable the named Pi-Hole."""
+            slug = slugify(name)
+            if slug not in hass.data[DOMAIN]:
+                raise vol.Invalid(
```

Comment:
A better solution would be to move the service schema inside `async_setup` and include the current valid names in the service schema validator, `vol.In`. Then we don't need this check.